International names support

AJenbo opened the pull request:
## Do not merge this as is, we need to figure out how to best handle things in a consistent way to how PHP works

Type: feature  
Breaking change: no

Filenames, namespaces and class names where being handled with functions that aren't safe on non anscii string, but PHP non-anscii names.

mb_* is already used in other parts of the application so this doesn't add new dependencies.

AJenbo commented:
The behavior of `strtolower` depends on the local setting, so `É` will be changed to `é` if your local is set to French. It also can corrupt UTF8 strings as it assumes the string is ISO/IEC 8859-1.
https://www.php.net/manual/en/function.strtolower.php

kylekatarnls commented on src/main/php/PDepend/Source/Language/PHP/PHPParserVersion71.php:
```diff
@@ -176,7 +176,7 @@ protected function isScalarOrCallableTypeHint($image)
     protected function parseScalarOrCallableTypeHint($image)
     {
-        switch (strtolower($image)) {
+        switch (mb_strtolower($image)) {
```
Not needed here: `void` and `iterable` has only ASCII characters.

kylekatarnls commented:
It would be way slower but the right move would be to do what PHP do (only as for as the class/function/interfaces/methods are concerned):

```php
preg_replace_callback('[a-zA-Z]', 'strtolower', $name)
```

This won't corrupt the string (can't return empty string like `strtolower` with non-UTF-8 charsets as default locale) and this would keep non-ASCII characters uppercased which is what we want as `é()` and `É()` should not be seen as the same function:
```php
function é() {
    return 42;
}

function É() {
    return 43;
}

var_dump(é(), É()); // 42 43
```

Which would happen if we merge this PR as is, such as in `hasExcludeFor()` or the symbol table. I have to check the tokenizer too.

Other places seem correct, except `parseScalarOrCallableTypeHint` which does not need `mb_strtolower` at all, so I suggest to keep the faster one: `strtolower` in this method.

AJenbo commented:
I would suggest always using the mb_ functions as it makes it much easier to see if the application has been proper I18n'ed or not. The performance would probably not be an issue and the behavior is more consistent.

A greedy match would probably also give you better performance (because it would match most strings and not just 1 char at a time):
```
preg_replace_callback('[a-zA-Z0-9_-]+', 'mb_strtolower', $name)
```

kylekatarnls commented:
> mb_* is already used in other parts of the application so this doesn't add new dependencies.

Actually no, we have only once and it's inside a `if (function_exists('mb_detect_encoding')) {}` and any of our non-dev dependencies use it.

So we need to add either the `"ext-mbstring"` in dependencies but it's actually a BC to consider or we can add a polyfill like the symfony one.

> see if the application has been proper I18n'ed or not

If we properly do, we can show it in unit tests and document it, we won't replace every substr with mb_substr everywhere if it does not add any value: fix a bug we can demonstrate with a unit test). Replacing every `strtolower` with `mb_strtolower` does not mean PDepend has been proper I18n'ed.

Every change is a possible regression we'll have to handle, so I want every of them to be weighted and only the useful ones to be kept.

I see you replaced `strtolower` with `mb_strtolower` and added `0-9_-` in `preg_replace_callback('[a-zA-Z0-9_-]+', 'mb_strtolower', $name)` so I think you don't get the point of this code.

It can be with the greed: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $name)` fine, but `mb_` does not make sense, we're only replacing ASCII characters a to z and A to Z: this will not include any non-ASCII letter. So `strtolower` is the right tool in the callback to use. And we don't need `0-9_-` to be proceeded by any lowercasing. For instance, you would not use `mb_` to lowercase `date('l')` you would use an ASCII replacement as you know the days of the week are only ASCII characters in English. And you could even only lowercase the first letter as you know only the first letter is uppercased: `lcfirst(date('l'))` the principle here is to parse and replace only what you need.

It must be considered wisely for each replacement. In some place, it does not change anything (like in `parseScalarOrCallableTypeHint` and `isScalarOrCallableTypeHint` which uses `strtolower` to compare with fully-ASCII predefined words, using `mb_strtolower` is asking the machine to proceed extra replacements we know we won't use. If there is any non-ASCII character, we want to go to the `default` case, and this would happen properly with any encoding in use using `strtolower` => `parseScalarOrCallableTypeHint` and `isScalarOrCallableTypeHint` fully support I18N) and in some other places, it changes the behavior, in this case we have to check if this behavior matches the given feature. And as per the example of `é()` and `É()` it does not match for function names. The purpose of the tool is to match the PHP compiler as much as possible when it comes to trying to determine which function/method/interface/class is used. If we say `é();` is calling the `function É(){}` declaration, this does not match the PHP behavior, this is not supporting I18N, this is simply a bug.

To summarize: what I would expect from a I18N support PR is to show in a unit test `Ñaming.php` fail to be proceeded by feature X or Y. So we change this `strtolower` in this particular method with `mb_strtolower` to make it work.

AJenbo commented:
> or we can add a polyfill like the symfony one

I believe it depends on `iconv` so it might result in a similar issue. Maybe it's best to wait until 3.0.

> Replacing every strtolower with mb_strtolower does not mean PDepend has been proper I18n'ed

Of cause not, but if we have to test every single part of the application to see how it deals UTF8 under various locals, it will be a very long time before getting to deal with any of the other steps in that process.

> I see you replaced strtolower with mb_strtolower and added 0-9_- in preg_replace_callback('[a-zA-Z0-9_-]+', 'mb_strtolower', $name) so I think you don't get the point of this code.

I think you are overthinking my example. I changed `mb_` as part of my previous point.

> And we don't need `0-9_-` to be proceeded by any lowercasing

It's there so that the match will be more likely to match the full name instead of calling `tolower` twice, for a name like `Core_PHP70Parser`. 0-9_ are commonly used and within the safe 127-byte chars. It also simplifies most of the matches, in the above case it takes it from 10 to 3 steps in the PCRE engine.

> The purpose of the tool is to match the PHP compiler as much as possible when it comes to trying to determine which function/method/interface/class is used. If we say é(); is calling the function É(){} declaration, this does not match the PHP behavior, this is not supporting I18N, this is simply a bug.

If you think I'm arguing anything else you have misunderstood me.

We know `strtolower` is inconsistent across installations and doesn't always match PHP's behavior. In my view having to evaluate between `strtolower` and  `mb_strtolower` on each case is a waste of everyone's time, the performance gain from using `strtolower` in some cases is most likely also negligible, and it would unnecessarily complicate the codebase. In short, I suggest we pick one function consistently (either force locale `C` and use `strtolower`, or use `mb_strtolower`) and then roll our own wrapper where/if we need to match PHP's behavior.

kylekatarnls commented:
`mb_strtolower` actually introduce I18N a bug (what I explained with É() and é()), you put too much trust in this function. And you still don't get the `preg_replace_callback` point:

```php
preg_replace_callback('[a-zA-Z]+', 'strtolower', 'A-5_É')
```
returns: `'a-5_É'` (it handle non-ASCII characters, digits, dashes, etc. the way we want: untouched) and it _works fine_ even with `setlocale(LC_ALL, 'fr.iso-8859-1')` because `strtolower` don't fail in any installation with `a-zA-Z` (ASCII characters).

You CAN handle I18N properly with `strtolower` and you can break features/compatibility using `mb_strtolower`. That's not a magical function that always return what is needed. That's why a blind replacements of everything is not an option. I won't be relieved at all by changing every `strtolower` with `mb_strtolower` and you shouldn't either.

And the point about not using `mb_` where it's not needed is not about performance, it's about using functions that make sense in the given context. You wouldn't do `mb_strtolower('ABC')`, here `mb` is overkill, `strtolower('ABC')` just do the job well and consistently across installations.

This one particular point can be discussed and probably other team members might have different points of view.

But about the `É()` / `é()` behavior, this must be handled with `strtolower` or `preg_replace_callback('[a-zA-Z]+', 'strtolower', $name)` not with `mb_strtolower` in order to avoid feature regression.

kylekatarnls commented on src/main/php/PDepend/Metrics/Analyzer/CodeRankAnalyzer/StrategyFactory.php:
```diff
@@ -110,7 +110,7 @@ public function createStrategy($strategyName)
 
         // Prepare identifier
-        $name = ucfirst(strtolower($strategyName));
+        $name = ucfirst(mb_strtolower($strategyName));
```
Useless, `->validStrategies` is a private array property which contains only ASCII strings (related to PHP ASCII keywords).

=> `$strategyName` will never contain non-ASCII characters here.

kylekatarnls commented on src/main/php/PDepend/Source/AST/ASTCastExpression.php:
```diff
@@ -89,7 +89,7 @@ class ASTCastExpression extends \PDepend\Source\AST\ASTUnaryExpression
     public function __construct($image)
     {
-        parent::__construct(preg_replace('(\s+)', '', strtolower($image)));
+        parent::__construct(preg_replace('(\s+)', '', mb_strtolower($image)));
```
`$image` here is supposed to be one of the following:

- (int), (integer) - cast to integer
- (bool), (boolean) - cast to boolean
- (float), (double), (real) - cast to float
- (string) - cast to string
- (array) - cast to array
- (object) - cast to object
- (unset) - cast to NULL

Should we handle invalid casts supposedly which could contain non-ASCII characters?

AJenbo replied:
The benefit of changing this case is that it would be clear that we applied the general solution and so not be unsure if this was an oversight or a special case. It also is a benefit to use the general case every where as it leaves less room for errors when code is refactored or copied.

kylekatarnls commented on src/main/php/PDepend/Source/AST/ASTTrait.php:
```diff
@@ -77,7 +77,7 @@ public function getAllMethods()
 
         foreach ($this->getMethods() as $method) {
-            $methods[strtolower($method->getName())] = $method;
+            $methods[mb_strtolower($method->getName())] = $method;
```
`é()` / `É()` regression

AJenbo replied:
@kylekatarnls please calm down, we have already identified this. Pointing it out over and over is totally unessesery. I don't know why you think I'm arguing for this!

kylekatarnls replied:
Putting it for each change where it causes the bug gives the overview of benefits vs. drawbacks.

AJenbo replied:
I think a single comment with a count would have done the job :|

kylekatarnls commented on src/main/php/PDepend/Source/AST/ASTTraitUseStatement.php:
```diff
@@ -90,13 +90,13 @@ public function getAllMethods()
     public function hasExcludeFor(ASTMethod $method)
     {
-        $methodName   = strtolower($method->getName());
+        $methodName   = mb_strtolower($method->getName());
```
`é()` / `É()` regression

kylekatarnls commented on src/main/php/PDepend/Source/AST/ASTTraitUseStatement.php:
```diff
@@ -90,13 +90,13 @@ public function getAllMethods()
 
         foreach ($precedences as $precedence) {
-            if (strtolower($precedence->getImage()) !== $methodName) {
+            if (mb_strtolower($precedence->getImage()) !== $methodName) {
```
`é()` / `É()` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $method->getName())` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented on src/main/php/PDepend/Source/AST/ASTTraitUseStatement.php:
```diff
@@ -139,11 +139,11 @@ private function collectMethods(ASTTraitReference $reference)
     private function getAliasesFor(ASTMethod $method)
     {
-        $name = strtolower($method->getName());
+        $name = mb_strtolower($method->getName());
```
`é()` / `É()` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $method->getName())` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented on src/main/php/PDepend/Source/AST/ASTTraitUseStatement.php:
```diff
@@ -139,11 +139,11 @@ private function collectMethods(ASTTraitReference $reference)
         $newNames = array();
         foreach ($this->getAliases() as $alias) {
-            $name2 = strtolower($alias->getImage());
+            $name2 = mb_strtolower($alias->getImage());
```
`é()` / `É()` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $method->getName())` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented on src/main/php/PDepend/Source/AST/AbstractASTClassOrInterface.php:
```diff
@@ -263,7 +263,7 @@ public function getAllMethods()
         foreach ($this->getInterfaces() as $interface) {
             foreach ($interface->getAllMethods() as $method) {
-                $methods[strtolower($method->getName())] = $method;
+                $methods[mb_strtolower($method->getName())] = $method;
```
`é()` / `É()` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $method->getName())` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented on src/main/php/PDepend/Source/AST/AbstractASTClassOrInterface.php:
```diff
@@ -274,11 +274,11 @@ public function getAllMethods()
 
         foreach ($this->getTraitMethods() as $method) {
-            $methods[strtolower($method->getName())] = $method;
+            $methods[mb_strtolower($method->getName())] = $method;
```
`é()` / `É()` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $method->getName())` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented on src/main/php/PDepend/Source/AST/AbstractASTClassOrInterface.php:
```diff
@@ -274,11 +274,11 @@ public function getAllMethods()
 
         foreach ($this->getMethods() as $method) {
-            $methods[strtolower($method->getName())] = $method;
+            $methods[mb_strtolower($method->getName())] = $method;
```
`é()` / `É()` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $method->getName())` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented on src/main/php/PDepend/Source/AST/AbstractASTType.php:
```diff
@@ -329,7 +329,7 @@ protected function getTraitMethods()
             /** @var ASTTraitAdaptationPrecedence $precedence */
             foreach ($precedences as $precedence) {
-                $priorMethods[strtolower($precedence->getImage())] = true;
+                $priorMethods[mb_strtolower($precedence->getImage())] = true;
```
`é()` / `É()` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $method->getName())` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented on src/main/php/PDepend/Source/AST/AbstractASTType.php:
```diff
@@ -339,7 +339,7 @@ protected function getTraitMethods()
                 }
 
-                $name = strtolower($method->getName());
+                $name = mb_strtolower($method->getName());
```
`é()` / `É()` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $method->getName())` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented on src/main/php/PDepend/Source/Language/PHP/PHPBuilder.php:
```diff
@@ -2100,7 +2100,7 @@ protected function findType(array $instances, $qualifiedName)
         $namespaceName = $this->extractNamespaceName($qualifiedName);
 
-        $caseInsensitiveName = strtolower($classOrInterfaceName);
+        $caseInsensitiveName = mb_strtolower($classOrInterfaceName);
```
`class é` / `class É` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $method->getName())` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented on src/main/php/PDepend/Source/Language/PHP/PHPBuilder.php:
```diff
@@ -2239,7 +2239,7 @@ public function restoreInterface(ASTInterface $interface)
     protected function storeTrait($traitName, $namespaceName, ASTTrait $trait)
     {
-        $traitName = strtolower($traitName);
+        $traitName = mb_strtolower($traitName);
```
`trait é` / `trait É` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $method->getName())` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented on src/main/php/PDepend/Source/Language/PHP/PHPBuilder.php:
```diff
@@ -2260,7 +2260,7 @@ protected function storeTrait($traitName, $namespaceName, ASTTrait $trait)
     protected function storeClass($className, $namespaceName, ASTClass $class)
     {
-        $className = strtolower($className);
+        $className = mb_strtolower($className);
```
`class é` / `class É` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $method->getName())` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented on src/main/php/PDepend/Source/Language/PHP/PHPBuilder.php:
```diff
@@ -2281,7 +2281,7 @@ protected function storeClass($className, $namespaceName, ASTClass $class)
     protected function storeInterface($interfaceName, $namespaceName, ASTInterface $interface)
     {
-        $interfaceName = strtolower($interfaceName);
+        $interfaceName = mb_strtolower($interfaceName);
```
`interface é` / `interface É` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', $method->getName())` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

AJenbo replied:
@kylekatarnls please stop spamming me.

AJenbo commented:
> mb_strtolower actually introduce I18N a bug (what I explained with É() and é()), you put too much trust in this function.

I'm not. We would filter the relevant through `preg_replace_callback` or similar, so it won't cause issues with `É` vs `é`. That's exactly what my example did. Doing so makes it irrelevant if `mb_strtolower` or `strtolower` is being used.

> That's why a blind replacements of everything is not an option.

That's not what I'm suggesting, despite you insinuating it.

> You wouldn't do mb_strtolower('ABC'), here mb is overkill

What is it that you feel is overkill about it?

kylekatarnls commented on src/main/php/PDepend/Source/Language/PHP/PHPParserVersion71.php:
```diff
@@ -159,7 +159,7 @@ protected function parseUnknownDeclaration($tokenType, $modifiers)
     protected function isScalarOrCallableTypeHint($image)
     {
-        switch (strtolower($image)) {
+        switch (mb_strtolower($image)) {
```
Useless: cases are fully-ASCII words.

kylekatarnls replied:
Please explain the resolution when you mark as resolve. We don't merge useless changes by default.

AJenbo replied:
Since we both agree that this PR has to be updated before it is ready to be merged it's pointless to make comments on each line at this stage.

I asked you to please stop commenting on the individual lines and spamming my inbox; marking them as resolved was the only way I could see to try and lower the noise.

Once we have come to an agreement on how to solve things in general we can revisit the individual lines after the code has been updated.

kylekatarnls commented on src/main/php/PDepend/Source/Language/PHP/PHPTokenizerInternal.php:
```diff
@@ -801,7 +801,7 @@ private function tokenize()
                 $startLine += $lines;
             } else {
-                $value = strtolower($token[1]);
+                $value = mb_strtolower($token[1]);
```
This one needs a unit test to demonstrate the before/after change.

kylekatarnls commented on src/main/php/PDepend/Source/Parser/SymbolTable.php:
```diff
@@ -170,6 +170,6 @@ private function ensureActiveScopeExists()
     private function normalizeKey($key)
     {
-        return strtolower($key);
+        return mb_strtolower($key);
```
This one needs a unit test to demonstrate the before/after change.

kylekatarnls commented on src/main/php/PDepend/Util/IdBuilder.php:
```diff
@@ -98,7 +98,7 @@ public function forClassOrInterface(AbstractASTType $type)
         return $this->forOffsetItem(
             $type,
-            ltrim(strrchr(strtolower(get_class($type)), '_'), '_')
+            ltrim(strrchr(mb_strtolower(get_class($type)), '_'), '_')
```
`class é` / `class É` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', get_class($type))` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented:
> irrelevant if mb_strtolower or strtolower is being used

That's not, so I will do the job of inspecting those changes one by one to show you in each case the possible issue.

kylekatarnls commented on src/main/php/PDepend/Util/IdBuilder.php:
```diff
@@ -112,7 +112,7 @@ public function forClassOrInterface(AbstractASTType $type)
     {
         $fileHash = $artifact->getCompilationUnit()->getId();
-        $itemHash = $this->hash($prefix . ':' . strtolower($artifact->getName()));
+        $itemHash = $this->hash($prefix . ':' . mb_strtolower($artifact->getName()));
```
Offsets are case sensitive, so the lowercasing here might be a mistake in the first place. We have to check what `$itemHash` is for.

The problem here might be than depending of what the artifact is, it shouldn't be lowercased. And when it's a class/method/function/interface/trait,`mb_strtolower` will cause the regression.

kylekatarnls commented on src/main/php/PDepend/Util/ImageConvert.php:
```diff
@@ -61,8 +61,8 @@ class ImageConvert
-        $inputType  = strtolower(pathinfo($input, PATHINFO_EXTENSION));
-        $outputType = strtolower(pathinfo($output, PATHINFO_EXTENSION));
+        $inputType  = mb_strtolower(pathinfo($input, PATHINFO_EXTENSION));
+        $outputType = mb_strtolower(pathinfo($output, PATHINFO_EXTENSION));
```
non-ASCII file extensions? I would encourage no one to do that. But it's correct to lowercase non-ASCII characters here if it would happen.

kylekatarnls commented on src/main/php/PDepend/Util/Type.php:
```diff
@@ -242,7 +242,7 @@ public static function isInternalType($typeName)
 
         $normalizedName = ltrim($typeName, '\\');
-        $normalizedName = strtolower($normalizedName);
+        $normalizedName = mb_strtolower($normalizedName);
```
Correct.

kylekatarnls commented on src/main/php/PDepend/Util/Type.php:
```diff
@@ -260,7 +260,7 @@ public static function getTypePackage($typeName)
 
         $normalizedName = ltrim($typeName, '\\');
-        $normalizedName = strtolower($normalizedName);
+        $normalizedName = mb_strtolower($normalizedName);
```
Correct.

kylekatarnls commented on src/main/php/PDepend/Util/Type.php:
```diff
@@ -294,7 +294,7 @@ public static function getInternalNamespaces()
     {
         $packageNames = self::getInternalNamespaces();
-        return isset($packageNames[strtolower($packageName)]);
+        return isset($packageNames[mb_strtolower($packageName)]);
```
Correct.

kylekatarnls commented on src/main/php/PDepend/Util/Type.php:
```diff
@@ -307,7 +307,7 @@ public static function isInternalPackage($packageName)
     public static function isScalarType($image)
     {
-        $image = strtolower($image);
+        $image = mb_strtolower($image);
```
Not needed, `$scalarTypes` is private and ASCII characters only.

kylekatarnls commented on src/main/php/PDepend/Util/Type.php:
```diff
@@ -343,7 +343,7 @@ public static function isPrimitiveType($image)
     public static function getPrimitiveType($image)
     {
-        $image = strtolower($image);
+        $image = mb_strtolower($image);
```
Irrelevant: `$primitiveTypes` is a private property and only ASCII characters.

kylekatarnls commented on src/main/php/PDepend/Util/Type.php:
```diff
@@ -369,7 +369,7 @@ public static function getPrimitiveType($image)
     public static function isArrayType($image)
     {
-        return (strtolower($image) === 'array');
+        return (mb_strtolower($image) === 'array');
```
Irrelevant: `'array'` is fully ASCII.

kylekatarnls commented on src/site/config.php:
```diff
@@ -115,7 +115,7 @@ public function relativeUrl($url)
                 $content = $match['content'];
                 // Use content as anchor
-                $hash = preg_replace('/[^a-z0-9]+/', '-', strtolower(trim($match['content'])));
+                $hash = preg_replace('/[^a-z0-9]+/', '-', mb_strtolower(trim($match['content'])));
```
Correct.

kylekatarnls commented on src/test/php/PDepend/AbstractTest.php:
```diff
@@ -809,7 +809,7 @@ public static function autoload($className)
     {
         $reflection = new \ReflectionClass('Iterator');
-        $extension  = strtolower($reflection->getExtensionName());
+        $extension  = mb_strtolower($reflection->getExtensionName());
```
Likely not to need non-ASCII support but correct.

kylekatarnls commented on src/test/php/PDepend/AbstractTest.php:
```diff
@@ -843,7 +843,7 @@ public function parseTestCaseSource($testCase, $ignoreAnnotations = false)
         list($class, $method) = explode('::', $testCase);
 
-        $fileName = substr(strtolower($class), 8, strrpos($class, '\\') - 8);
+        $fileName = substr(mb_strtolower($class), 8, strrpos($class, '\\') - 8);
```
Irrelevant: we have any non-ASCII characters in our tes files names.

kylekatarnls commented on src/test/php/PDepend/Util/IdBuilderTest.php:
```diff
@@ -95,7 +95,7 @@ public function testBuilderCreatesCaseSensitiveFileIdentifiers()
                 ->getMock();
         $unitStub1->method('getFileName')
-                ->willReturn(strtolower(__FILE__));
+                ->willReturn(mb_strtolower(__FILE__));
```
Not used yet but correct.

kylekatarnls commented on src/test/php/PDepend/Util/IdBuilderTest.php:
```diff
@@ -173,7 +173,7 @@ public function testBuilderCreatesCaseInSensitiveClassIdentifiers()
         $class0->setCompilationUnit($compilationUnit);
         
-        $class1 = new ASTClass(strtolower(__FUNCTION__));
+        $class1 = new ASTClass(mb_strtolower(__FUNCTION__));
```
`é()` / `É()` regression: `preg_replace_callback('[a-zA-Z]+', 'strtolower', __FUNCTION__)` can be use as a replacement to avoid both the regression and the charset PHP malfunction when lowercasing non-ASCII characters with `strtolower`.

kylekatarnls commented on src/test/php/PDepend/Util/IdBuilderTest.php:
```diff
@@ -198,7 +198,7 @@ public function testBuilderCreatesCaseInSensitiveInterfaceIdentifiers()
         $interface0->setCompilationUnit($compilationUnit);
 
-        $interface1 = new ASTInterface(strtolower(__FUNCTION__));
+        $interface1 = new ASTInterface(mb_strtolower(__FUNCTION__));
```
Lowercasing here might be a bug in the first place.

kylekatarnls commented on src/test/php/PDepend/Util/IdBuilderTest.php:
```diff
@@ -241,7 +241,7 @@ public function testBuilderCreatesCaseInSensitiveFunctionIdentifiers()
         $function0->setCompilationUnit($compilationUnit);
 
-        $function1 = new ASTFunction(strtolower(__FUNCTION__));
+        $function1 = new ASTFunction(mb_strtolower(__FUNCTION__));
```
Lowercasing here might be a bug in the first place.

kylekatarnls commented on src/test/php/PDepend/Util/IdBuilderTest.php:
```diff
@@ -328,7 +328,7 @@ public function testBuilderCreatesCaseInSensitiveMethodIdentifiers()
         $method0->setParent($class);
 
-        $method1 = new ASTMethod(strtolower(__FUNCTION__));
+        $method1 = new ASTMethod(mb_strtolower(__FUNCTION__));
```
Lowercasing here might be a bug in the first place.

kylekatarnls commented:
I will check what are the $class0 / $class1 and other double objects like this in our tests utils. I think we could completely get rid of case-insensitivity here.

AJenbo commented:
> > irrelevant if mb_strtolower or strtolower is being used
> 
> That's not, so I will do the job of inspecting those changes one by one to show you in each case the possible issue.

I feel like you are taking me out of context. We both seem to agree that we should use `preg_replace_callback` to make sure only the relevant part is converted (something you omitted from the quote). Do you not agree that `preg_replace_callback('[a-zA-Z]+', 'strtolower', $name)` and `preg_replace_callback('[a-zA-Z]+', 'mb_strtolower', $name)` would produce the exact same result in all cases?

Like you commented on several lines here (Irrelevant/Useless/Not needed) there is no difference to using one over the other.

If it makes no difference if `strtolower` or `mb_strtolower` is used, I don't see how you can say that one is more right for the job then the other?

kylekatarnls commented:
> preg_replace_callback('[a-zA-Z]+', 'strtolower', $name) and preg_replace_callback('[a-zA-Z]+', 'mb_strtolower', $name) would produce the exact same result in all cases

Correct. So if I have to choose, I take the shorter and more compatible (and faster) one.

> If it makes no difference if strtolower or mb_strtolower is used, I don't see how you can say that one is more right for the job then the other?

If it makes no difference, I vote for the *statu quo* => not changing the code.

AJenbo commented:
Ok, I think we are on the same page now :)

> If it makes no difference, I vote for the statu quo => not changing the existing code.

My personal preference is for consistency over specialization. Since we don't agree I will wait and see what the general opinion in the maintainer team is and then adjust my PR accordingly.

ravage84 commented:
To be honest with you @AJenbo, I'm _not_ in favor of this. Using non-ASCII characters in programming is a potential mess if not a call for trouble.

And @kylekatarnls seems to point out potential problems arising from introducing this support.

Also, this hasn't been much of an issue, it seems. As there are no [past issues on ASCII](https://github.com/pdepend/pdepend/issues?q=is%3Aissue+is%3Aopen+ascii) or [international names](https://github.com/pdepend/pdepend/issues?q=is%3Aissue+is%3Aopen+international) and only one about [umlauts](https://github.com/pdepend/pdepend/issues?q=is%3Aissue+umlaut+is%3Aclosed).

> So we need to add either the `"ext-mbstring"` in dependencies but it's actually a BC to consider or we can add a polyfill like the symfony one.

@kylekatarnls adding a dependency isn't a BC break. Though one should be cautious with it, I'd agree.

kylekatarnls commented:
> is a potential mess

PHPMD can judge this. But I think PDepend should not take this into consideration and should aim to parse valid code, not necessarily clean code.

That's why a I still think some of the changes here are good (but only those which are useful and regression-safe - this is for my vote) and the benefit of those changes could be demonstrated by some unit tests. It would help to show where we support non-ASCII characters.

ravage84 commented:
I'm only OK with this if we greatly improve the code coverage concerning potential naming cases.

Also, I would prefer to split this PR into several small ones. Each introducing new tests that prove the code changes do not introduce bugs.

ravage84 commented:
Closing this for now. We might come back at a later time, if we can assure the functionality through tests.

Thank you for your efforts, @AJenbo
